lib/pdf/conversion.rb

```diff
@@ -23,13 +23,17 @@ def self.generate_from_html(html, outpath, options = {})
 
   PDF_GENERATION_TIMEOUT = 15 * 60
 
-  def self.invoke_generation_script(args, options = {})
+  def self.invoke_generation_script(args, options = {}, attempts: 3)
     script_path = "#{deploy_dir}/bin/generate-pdf"
     cmd = (['timeout', PDF_GENERATION_TIMEOUT.to_s, 'node', script_path] + args).join(" ")
     puts cmd if options[:verbose]
     @result = `#{cmd}`
     if $?.exitstatus == 124
       raise "pdf generation timed out after #{PDF_GENERATION_TIMEOUT} seconds. cmd: #{cmd}"
+    elsif $?.exitstatus == 1
+      warn "Exit status 1.  Re-attempting."
+      raise "pdf generation failed with status #{$?.exitstatus}. cmd: #{cmd}" if attempts <= 1
+      invoke_generation_script(args, options, attempts: attempts - 1)
```

Comment:
Recursion always sets off my code smell detectors. This works but I'm curious why not simply use a loop? To me, that's generally more readable, but I'm open to discussion

Reply:
Good question - I was leveraging [this](https://github.com/code-dot-org/code-dot-org/blob/fc372301b0b1a4543fe5560b9f91d20afb5c8338/lib/cdo/crowdin/project.rb#L42) and [this](https://github.com/code-dot-org/code-dot-org/blob/fc372301b0b1a4543fe5560b9f91d20afb5c8338/lib/cdo/crowdin/legacy_utils.rb#L88) as examples. 

That being said, a loop seems like an equally good solution here.

Reply:
If we used a loop, could we include the timeout above in the retry logic? Do we want to?

If the answer to both those questions is yes, then I think this should be turned into a loop. If the answer to one (or both) of those questions is no, I think this is fine to avoid a refactor / increased complexity outside of this change. This recursion is straightforward and not going to have a performance impact, so it's not something I'm going to block on. :)

Reply:
I did try changing this to a loop inside the `elsif $?.exitstatus == 1` case, it looked something like this:

```
def self.invoke_generation_script(args, options = {})
    script_path = "#{deploy_dir}/bin/generate-pdf"
    cmd = (['timeout', PDF_GENERATION_TIMEOUT.to_s, 'node', script_path] + args).join(" ")
    puts cmd if options[:verbose]
    @result = `#{cmd}`
    if $?.exitstatus == 124
      raise "pdf generation timed out after #{PDF_GENERATION_TIMEOUT} seconds. cmd: #{cmd}"
    elsif $?.exitstatus == 1
      retryAttempts = 2
      while $?.exitStatus == 1 && retryAttempts > 0
        retryAttempts = retryAttempts - 1
        warn "Exit status 1.  Re-attempting."
        cmd = (['timeout', PDF_GENERATION_TIMEOUT.to_s, 'node', script_path] + args).join(" ")
        puts cmd if options[:verbose]
        @result = `#{cmd}`
      end
    elsif $?.exitstatus != 0
      raise "pdf generation failed with status #{$?.exitstatus}. cmd: #{cmd}"
    end
    puts "  #{@result}" if options[:verbose]
  end
```
but I ran into an issue that if the re-attempt failed for a different reason, that reason wouldn't be captured in the output.  

That led me to try wrapping nearly all of the if-else logic inside a while loop but it got pretty ugly looking by the end of that... so I changed it back to use recursion.  

That being said, I am not so sure how important it is to raise all the different exit statuses in this function.  It is possible that we just say "if you fail once, try 2 more times" and then raise the last failure that took place.   I also might be totally overthinking this.

Thoughts?

Reply:
Yeah, I suppose it is a bit awkward to implement a looped retry for just that one failure case. If we instead wanted to just always retry, we could do something like:

```ruby
def self.invoke_generation_script(args, options = {}, retry_attempts: 3)
  script_path = "#{deploy_dir}/bin/generate-pdf"
  cmd = (['timeout', PDF_GENERATION_TIMEOUT.to_s, 'node', script_path] + args).join(" ")
  puts cmd if options[:verbose]
  current_attempt = 0
  while current_attempt < retry_attempts
    warn "Retrying; attempt #{current_attempt}" if current_attempt > 0
    current_attempt += 1
    @result = `#{cmd}`
    if $?.exitstatus == 124 # Timeout
      message = "pdf generation timed out after #{PDF_GENERATION_TIMEOUT} seconds. cmd: #{cmd}"
    elsif $?.exitstatus != 0 # Any other non-success
      message = "pdf generation failed with status #{$?.exitstatus}. cmd: #{cmd}"
    elsif $?.exitstatus == 0 # Success!
      break
    end

    if current_attempt < retry_attempts
      warn message
    else
      raise message
    end
  end
  puts "  #{@result}" if options[:verbose]
end
```

But that's starting to get a bit silly and I think your first implementation is probably cleaner.

What about this plan: keep your original recursive implementation, but add a comment which calls out that we are intentionally using recursion here even though it is often a code smell, and explains that we are doing so because in this specific case we are confident that performance is not a concern and we found that our attempts to implement this with a loop just led to even worse-smelling code. Just a little note to whatever poor soul finds themselves trying to iterate on this method in the future :smile: